feat: add `CommunityTokenPermission` to community description

0x-r4bbit opened the pull request:
This adds token permission capabilities to `CommunityDescription` such that token permissions can be created, updated and deleted.

ghost commented:
<!--prchecklist--> 
### Pull Request Checklist
- [ ] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://status.im/docs/))?
- [ ] Have you tested changes with mobile?
- [ ] Have you tested changes with desktop?

0x-r4bbit commented on protocol/protobuf/communities.proto:
```diff
@@ -36,6 +36,37 @@ message CommunityPermissions {
+  enum Type {
+    UNKNOWN_TOKEN_PERMISSION = 0;
+    BECOME_ADMIN = 1;
+    BECOME_MEMBER = 2;
```
This will have two more types, but not planned for MVP

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (37)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :x: | 963794a6 | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3207/1/) | 2023-02-20 16:38:24 | ~30 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3207/1/consoleText) |
| :heavy_check_mark: | 963794a6 | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3207/1/) | 2023-02-20 16:39:36 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230220-163758-963794-pr3207.zip) |
| :heavy_check_mark: | 963794a6 | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3207/1/) | 2023-02-20 16:41:59 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230220-163753-963794-pr3207.aar) |
| :heavy_check_mark: | 963794a6 | [#1](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3207/1/) | 2023-02-20 16:48:29 | ~10 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3207/1//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 1627e074 | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3207/2/) | 2023-02-22 13:49:46 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230222-134827-1627e0-pr3207.zip) |
| :heavy_check_mark: | 1627e074 | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3207/2/) | 2023-02-22 13:52:12 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230222-134827-1627e0-pr3207.zip) |
| :heavy_multiplication_x: | 1627e074 | [#2](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3207/2/) | 2023-02-22 13:52:12 | ~3 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3207/2//consoleText) |
| :heavy_check_mark: | 1627e074 | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3207/2/) | 2023-02-22 13:53:00 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230222-134825-1627e0-pr3207.aar) |
| | | | | | | |
| :heavy_check_mark: | cfd1b1fd | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3207/3/) | 2023-02-24 08:08:59 | ~55 sec | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230224-080803-cfd1b1-pr3207.zip) |
| :heavy_check_mark: | cfd1b1fd | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3207/3/) | 2023-02-24 08:11:29 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230224-080803-cfd1b1-pr3207.aar) |
| :heavy_check_mark: | cfd1b1fd | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3207/3/) | 2023-02-24 08:11:56 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230224-080803-cfd1b1-pr3207.zip) |
| :heavy_check_mark: | cfd1b1fd | [#3](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3207/3/) | 2023-02-24 08:18:32 | ~10 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3207/3//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 8acb4994 | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3207/4/) | 2023-02-27 09:26:19 | ~5 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230227-092104-8acb49-pr3207.zip) |
| :heavy_check_mark: | 8acb4994 | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3207/4/) | 2023-02-27 09:29:47 | ~8 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230227-092104-8acb49-pr3207.aar) |
| :heavy_multiplication_x: | 8acb4994 | [#4](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3207/4/) | 2023-02-27 09:30:11 | ~9 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3207/4//consoleText) |
| :heavy_check_mark: | 8acb4994 | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3207/4/) | 2023-02-27 09:35:23 | ~14 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230227-092104-8acb49-pr3207.zip) |
| | | | | | | |
| :heavy_check_mark: | d2e867f6 | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3207/5/) | 2023-02-27 14:36:45 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230227-143522-d2e867-pr3207.zip) |
| :heavy_check_mark: | d2e867f6 | [#5](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3207/5/) | 2023-02-27 14:38:20 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230227-143522-d2e867-pr3207.zip) |
| :heavy_check_mark: | d2e867f6 | [#5](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3207/5/) | 2023-02-27 14:39:37 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230227-143518-d2e867-pr3207.aar) |
| :heavy_check_mark: | d2e867f6 | [#5](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3207/5/) | 2023-02-27 14:45:07 | ~9 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3207/5//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 19b3c832 | [#6](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3207/6/) | 2023-02-28 15:04:22 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230228-150317-19b3c8-pr3207.zip) |
| :heavy_check_mark: | 19b3c832 | [#6](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3207/6/) | 2023-02-28 15:06:37 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230228-150317-19b3c8-pr3207.zip) |
| :heavy_check_mark: | 19b3c832 | [#6](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3207/6/) | 2023-02-28 15:07:02 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230228-150317-19b3c8-pr3207.aar) |
| :heavy_check_mark: | 19b3c832 | [#6](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3207/6/) | 2023-02-28 15:14:30 | ~11 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3207/6//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 0cbf4a98 | [#7](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3207/7/) | 2023-03-01 12:22:06 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230301-121927-0cbf4a-pr3207.zip) |
| :heavy_check_mark: | 0cbf4a98 | [#7](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3207/7/) | 2023-03-01 12:25:20 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230301-121926-0cbf4a-pr3207.aar) |
| :heavy_check_mark: | 0cbf4a98 | [#7](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3207/7/) | 2023-03-01 12:30:47 | ~11 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230301-121927-0cbf4a-pr3207.zip) |
| :heavy_check_mark: | 0cbf4a98 | [#7](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3207/7/) | 2023-03-01 12:34:00 | ~14 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3207/7//consoleText) |
| | | | | | | |
| :heavy_check_mark: | e8e62242 | [#8](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3207/8/) | 2023-03-02 10:47:47 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230302-104620-e8e622-pr3207.zip) |
| :heavy_multiplication_x: | e8e62242 | [#8](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3207/8/) | 2023-03-02 10:48:59 | ~2 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3207/8//consoleText) |
| :heavy_check_mark: | e8e62242 | [#8](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3207/8/) | 2023-03-02 10:49:45 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230302-104616-e8e622-pr3207.zip) |
| :heavy_check_mark: | e8e62242 | [#8](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3207/8/) | 2023-03-02 10:50:53 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230302-104616-e8e622-pr3207.aar) |
| :heavy_check_mark: | e8e62242 | [#9](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3207/9/) | 2023-03-02 15:23:37 | ~10 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3207/9//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 345255ec | [#9](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3207/9/) | 2023-03-02 16:40:53 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230302-163839-345255-pr3207.zip) |
| :heavy_check_mark: | 345255ec | [#9](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3207/9/) | 2023-03-02 16:44:44 | ~6 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230302-163839-345255-pr3207.aar) |
| :heavy_multiplication_x: | 345255ec | [#10](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3207/10/) | 2023-03-02 16:46:31 | ~7 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3207/10//consoleText) |
| :heavy_check_mark: | 345255ec | [#9](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3207/9/) | 2023-03-02 16:46:33 | ~7 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230302-163839-345255-pr3207.zip) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 1da0aaa7 | [#10](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3207/10/) | 2023-03-02 17:13:20 | ~4 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230302-170855-1da0aa-pr3207.zip) |
| :heavy_check_mark: | 1da0aaa7 | [#10](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3207/10/) | 2023-03-02 17:18:38 | ~9 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230302-170855-1da0aa-pr3207.aar) |
| :heavy_check_mark: | 1da0aaa7 | [#11](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3207/11/) | 2023-03-02 17:23:57 | ~15 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3207/11//consoleText) |
| :heavy_check_mark: | 1da0aaa7 | [#10](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3207/10/) | 2023-03-02 17:32:17 | ~23 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230302-170855-1da0aa-pr3207.zip) |
| | | | | | | |
| :heavy_check_mark: | 4839b5a3 | [#11](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-3207/11/) | 2023-03-03 07:25:33 | ~57 sec | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-230303-072435-4839b5-pr3207.zip) |
| :heavy_check_mark: | 4839b5a3 | [#11](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-3207/11/) | 2023-03-03 07:27:24 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-230303-072435-4839b5-pr3207.zip) |
| :heavy_check_mark: | 4839b5a3 | [#11](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-3207/11/) | 2023-03-03 07:28:30 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-230303-072430-4839b5-pr3207.aar) |
| :heavy_multiplication_x: | 4839b5a3 | [#12](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3207/12/) | 2023-03-03 07:31:49 | ~7 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3207/12//consoleText) |
| :heavy_multiplication_x: | 4839b5a3 | [#13](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3207/13/) | 2023-03-03 08:03:11 | ~2 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3207/13//consoleText) |
| :heavy_check_mark: | 4839b5a3 | [#14](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3207/14/) | 2023-03-03 08:20:30 | ~8 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-3207/14//consoleText) |

0x-r4bbit commented on protocol/protobuf/communities.proto:
```diff
@@ -36,6 +36,37 @@ message CommunityPermissions {
+  string id = 1;
+  Type type = 2;
+  repeated TokenCriteria token_criteria = 3;
+  repeated string chat_ids = 4;
```
`chat_ids` are also ignored for MVP

jrainville commented on protocol/communities/community.go:
```diff
@@ -1390,6 +1399,80 @@ func (o *Community) Categories() map[string]*protobuf.CommunityCategory {
+		o.config.CommunityDescription.TokenPermissions = make(map[string]*protobuf.CommunityTokenPermission)
+	}
+
+	if _, ok := o.config.CommunityDescription.TokenPermissions[permission.Id]; ok {
```
nitpick, but maybe `exists` would be a better name than `ok`

jrainville commented on protocol/communities/community.go:
```diff
@@ -1390,6 +1399,80 @@ func (o *Community) Categories() map[string]*protobuf.CommunityCategory {
+	o.mutex.Lock()
+	defer o.mutex.Unlock()
+
+	if o.config.PrivateKey == nil {
```
Is this the best way to check if you're admin? I wonder if when we have owners and admins, does only the owner have the privatekey? Can the admin add permissions anyway? If no, then disregard this comment.

0x-r4bbit replied:
This is a good point. As per design, it looks like admins can create a subset of permissions as well.
But I see a problem here. 

Let's assume you don't need to own the private key here to make changes to the community description, but merely have the admin role. We can build it so that if you don't have the role, we'll exit here, but nothing can prevent you from actually changing the source, so that you **can** change the community description.

It then boils down to whether or not the updated community description was signed by the community (its private key), otherwise everyone could just make changes to a community.

So I guess, we then have to check what happens if:

1. We remove this check and we try to publish the community description without the private key being available
2. If there's anything in place that allows for 1), then we should check whether any community related messages are either signed by the community **or** by someone who has admin roles.

0x-r4bbit replied:
Okay, so I did some digging and it turns out, basically any changes made that haven't been signed by the community itself are being ignored:

https://github.com/status-im/status-go/blob/d0cc036d486092262382d45967db11f19d66641e/protocol/communities/community.go#L958-L960

So I believe we need to extend the protocol to make this work for admins.
Essentially something like: 

- If a community description message is being sent and the private key of that community is not available, check if the current account is a) a member and b) has admin roles. If a) and b) is the case, then the message will be signed by the admin's account
- If we receive a community description and the message wasn't signed by the community itself, check if it was signed by an admin member

However, as I'm writing this out I realized, this is probably not a good path forward either.
If we implement this, then this will enable "admins" to do anything they want really, because access to changing community description is an all or nothing game. So they themselves could make others admins etc.

Just thinking out loud here, but this smells like our protocol won't easily support this.

@cammellos summoning you for you additional valuable thoughts on this.

0x-r4bbit replied:
Discussed this with @cammellos offline and right now there's no solution for this in place and needs further exploratoin/research.

So I'll continue assuming that only owners can create permissions for now until we have a solution for this.

jrainville commented on protocol/requests/create_community_token_permission_request.go:
```diff
@@ -0,0 +1,56 @@
+
+var (
+	ErrCreateCommunityTokenPermissionInvalidCommunityID    = errors.New("create community token permission needs a valid community id")
+	ErrCreateCommunityTokenPermissionTooManyTokenCriteria  = errors.New("too many token criertia")
```
```suggestion
	ErrCreateCommunityTokenPermissionTooManyTokenCriteria  = errors.New("too many token criteria")
```

jrainville commented on protocol/requests/create_community_token_permission_request.go:
```diff
@@ -0,0 +1,56 @@
+		return ErrCreateCommunityTokenPermissionInvalidCommunityID
+	}
+
+	if len(p.TokenCriteria) > 5 {
```
Should we move this to a constant somewhere so we can modify it only in one place in case we need this type of check somewhere else?

0x-r4bbit replied:
Yes, I'll make it a constant

jrainville reviewed (commented):
Great job! LGTM, just 4 small comments. Only the question about admins is "important"

0x-r4bbit commented:
@jrainville this is ready for re-review

jrainville reviewed (approved):
Nice work

0x-r4bbit committed:
feat: add CommunityTokenPermission to community description

This adds token permission capabilities to CommunityDescription such that token permissions can be created, updated and deleted.
